File: leetcode/binary_search/302_Smallest_Rectangle_Enclosing_Black_Pixels/codes/bs.py
class Solution:
    def minArea(self, image: list[list[str]], x: int, y: int) -> int:
        m, n = len(image), len(image[0])

        def is_col_empty(j):
            return all(image[i][j] == '0' for i in range(m))

        def is_row_empty(i):
            return all(image[i][j] == '0' for j in range(n))

        def binarySearch(low, high, is_empty, search_min):
            while low < high:
                mid = (low + high) // 2
                if is_empty(mid):
                    if search_min:
                        low = mid + 1 
                    else:
                        high = mid   
                else:
                    if search_min:
                        high = mid   
                    else:
                        low = mid + 1  
            return low

        left = binarySearch(0, y, is_col_empty, True)
        right = binarySearch(y + 1, n, is_col_empty, False)
        top = binarySearch(0, x, is_row_empty, True)
        bottom = binarySearch(x + 1, m, is_row_empty, False)

        return (right - left) * (bottom - top)

File: leetcode/binary_search/302_Smallest_Rectangle_Enclosing_Black_Pixels/codes/bs.c
#include <stdbool.h>

bool isColEmpty(char** image, int m, int col) {
    for (int i = 0; i < m; ++i) {
        if (image[i][col] == '1') return false;
    }
    return true;
}

bool isRowEmpty(char** image, int n, int row) {
    for (int j = 0; j < n; ++j) {
        if (image[row][j] == '1') return false;
    }
    return true;
}

int binarySearch(char** image, int low, int high, bool (*isEmpty)(char**, int, int), int dim, bool searchMin) {
    while (low < high) {
        int mid = (low + high) / 2;
        if (isEmpty(image, dim, mid) == searchMin)
            low = mid + 1;
        else
            high = mid;
    }
    return low;
}

int minArea(char** image, int imageSize, int* imageColSize, int x, int y) {
    int m = imageSize, n = imageColSize[0];

    int left = binarySearch(image, 0, y, isColEmpty, m, true);
    int right = binarySearch(image, y + 1, n, isColEmpty, m, false);
    int top = binarySearch(image, 0, x, isRowEmpty, n, true);
    int bottom = binarySearch(image, x + 1, m, isRowEmpty, n, false);

    return (right - left) * (bottom - top);
}
